src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+    unsigned opcode        = hash_opcode(ins);
+    encode_info encode_val = encode_func(ins, instr2_value, opcode);
+    if(!encode_val.is_valid())
+        return nullptr;
+
+    key_type key            = encode_val.get_key();
+    hash_value_ptr hash_val = nullptr;
+
+    if(encode2_value.find(key) != encode2_value.end())
+    {
+        hash_val = encode2_value[key];
+        add_instr(hash_val->id);
+        instr2_value[ins] = hash_val;
+    }
+    else
+    {
+        hash_val           = &(create_value(ins));
+        encode2_value[key] = hash_val;
+    }
+    for(auto&& input : encode_val.get_inputs())
+    {
+        add_input(hash_val->id, input);
+        add_output(input->id, hash_val);
+    }
+    return hash_val;
+}
+
+hash_value& horizontal_fusion_impl::create_value(instruction_ref ins)
+{
+    unsigned id = static_cast<unsigned>(values.size());
+    values.push_back(hash_value{id, cur_point});
+    hash_value& val = get_value(id);
+    add_instr(id);
+    instr2_value[ins] = &val;
+    return val;
+}
+
+void horizontal_fusion_impl::process(instruction_ref ins)
+{
+    // Do not hash literals.
+    if(ins->name() == "@literal")
+        return;
+    if(instr2_hash.find(ins) != instr2_hash.end())
+    {
+        // Hash this instruction.
+        if(hash(ins) != nullptr)
+        {
+            for(auto&& output : ins->outputs())
+            {
+                instr2_hash[output] = true;
+            }
+            return;
+        }
+    }
+
+    std::unordered_map<std::string, int> op2_cnt;
+    bool hash_child = false;
+    // Do hash if at least two outputs have same operations.
+    for(auto&& output : ins->outputs())
+    {
+        const std::string& str = output->name();
+        if(op2_cnt.find(str) == op2_cnt.end())
+            op2_cnt[str] = 1;
+        else
+        {
+            op2_cnt[str] += 1;
+            hash_child = true;
+            break;
+        }
+    }
+    if(hash_child)
+    {
+        // Create a value for this instruction.
+        hash_value& value = create_value(ins);
+        value.set_root();
+        // Flag children to be hashed.
+        for(auto&& output : ins->outputs())
+        {
+            if(op2_cnt[output->name()] > 1)
+                instr2_hash[output] = true;
+        }
+    }
+}
+
+// Find the first axis that matches given dim.
+int horizontal_fusion_impl::find_axis(instruction_ref ins, int dim)
+{
+    int ndx = 0;
+    for(auto&& size : ins->get_shape().lens())
+    {
+        if(size == dim)
+            return ndx;
+        ndx++;
+    }
+    return -1;
+}
+
+// Find axis for convolution filter's output and broadcast's input.
+int horizontal_fusion_impl::find_axis(instruction_ref ins, instruction_ref base, int base_axis)
+{
+    if(is_conv(base))
+    {
+        return get_conv_output_axis();
+    }
+    else if(ins->outputs().at(0)->name() == "broadcast")
+    {
+        if(!match_dim(base, ins->outputs().at(0), -1))
```

Comment:
This should pass `ins->outputs().at(0)->get_shape().lens().size()` not `-1`.

Reply:
Any number that lies outside [0, ins->outputs().at(0)->get_shape().lens().size() -1 ] will do.
Fixed anyway.

Reply:
It denotes all elements.  You can use many numbers.  But fixed anyway.